src/it/scala/io/iohk/ethereum/db/IodbDataSourceIntegrationSuite.scala

```diff
@@ -0,0 +1,148 @@
+      val db = updateInSeparateCalls(
+        dataSource = new IodbDataSource(new LSMStore(dir = dir, keySize = KeySize)),
+        toUpsert = keyList.zip(keyList)
+      )
+      keyList.foreach { key => assert(db.get(OtherNamespace, key).contains(key)) }
+    }
+  }
+
+  test("IodbDataSource insert keys in a single update"){
+    forAll(seqByteStringOfNItemsGen(KeySizeWithoutPrefix)) { unFilteredKeyList: Seq[ByteString] =>
+      //create temporary dir
+      val dir = File.createTempFile("iodbInsert", "iodbInsert")
+      dir.delete()
+      dir.mkdir()
+
+      val keyList = unFilteredKeyList.take(KeyNumberLimit)
+      val db = new IodbDataSource(new LSMStore(dir = dir, keySize = KeySize))
+        .update(OtherNamespace, Seq(), keyList.zip(keyList))
+
+      keyList.foreach { key => assert(db.get(OtherNamespace, key).contains(key)) }
+    }
+  }
+
+  test("IodbDataSource update keys in separate updates"){
+    forAll(seqByteStringOfNItemsGen(KeySizeWithoutPrefix)) { unFilteredKeyList: Seq[ByteString] =>
+      //create temporary dir
+      val dir = File.createTempFile("iodbUpdate", "iodbUpdate")
+      dir.delete()
+      dir.mkdir()
+
+      val keyList = unFilteredKeyList.take(KeyNumberLimit)
+      val db = new IodbDataSource(new LSMStore(dir = dir, keySize = KeySize))
+        .update(OtherNamespace, Seq(), keyList.zip(keyList))
+
+      val keyListWithExtraByte = keyList.map(1.toByte +: _)
+      val dbAfterUpdate = updateInSeparateCalls(db, keyList.zip(keyListWithExtraByte))
+
+      keyList.zip(keyListWithExtraByte).foreach { case (key, value) =>
+        assert(dbAfterUpdate.get(OtherNamespace, key).contains(value)) }
+    }
+  }
+
+  test("IodbDataSource update keys in a single update"){
+    forAll(seqByteStringOfNItemsGen(KeySizeWithoutPrefix)) { unFilteredKeyList: Seq[ByteString] =>
+      //create temporary dir
+      val dir = File.createTempFile("iodbUpdate", "iodbUpdate")
+      dir.delete()
+      dir.mkdir()
+
+      val keyList = unFilteredKeyList.take(KeyNumberLimit)
+      val db = new IodbDataSource(new LSMStore(dir = dir, keySize = KeySize))
+        .update(OtherNamespace, Seq(), keyList.zip(keyList))
+
+      val keyListWithExtraByte = keyList.map(1.toByte +: _)
+      val dbAfterUpdate = db.update(OtherNamespace, Seq(), keyList.zip(keyListWithExtraByte))
+
+      keyList.zip(keyListWithExtraByte).foreach { case (key, value) =>
+        assert(dbAfterUpdate.get(OtherNamespace, key).contains(value))
+      }
+    }
+  }
+
+  test("IodbDataSource insert/update with invalid length") {
+    forAll(seqByteStringOfNItemsGen(KeySizeWithoutPrefix)) { unFilteredKeyList: Seq[ByteString] =>
+      //create temporary dir
+      val dir = File.createTempFile("iodbInvalidLength", "iodbInvalidLength")
+      dir.delete()
+      dir.mkdir()
+
+      val keyList = unFilteredKeyList.take(KeyNumberLimit)
+      val (keysLeft, keysToInsert) = Random.shuffle(keyList).splitAt(Gen.choose(0, keyList.size/2).sample.get)
```

Comment:
Why `Random.shuffle`? `unFilteredKeyList` is already randomized.
Apart from that, in case of failure this test will be hard to reproduce.